Rust/src/random_cut_forest.rs

```diff
@@ -0,0 +1,250 @@
+/// ```
+pub struct RandomCutForest<T> {
+    dimension: usize,
+    num_observations: usize,
+    num_trees: usize,
+    sample_size: usize,
+    time_decay: f32,
+    trees: Vec<SampledTree<T>>,
+}
+
+impl<T> RandomCutForest<T>
+    where T: Float + Sum
+{
+
+    /// Update a random cut forest with a new data point.
+    ///
+    /// A copy of the data point will be sent to each sampled tree in the forest
+    /// for consideration. Each tree independently decides whether to accept the
+    /// point into its sample.
+    ///
+    /// # Panics
+    ///
+    /// If the dimensionality of the input data point does not match the
+    /// dimensionality of the forest.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use random_cut_forest::{RandomCutForest, RandomCutForestBuilder};
+    ///
+    /// // create a default RCF on two-dimensional data
+    /// let mut forest: RandomCutForest<f32> = RandomCutForestBuilder::new(2).build();
+    ///
+    /// // update the forest with some data
+    /// forest.update(vec![0.0, 0.0]);
+    /// forest.update(vec![1.0, 0.0]);
+    /// forest.update(vec![0.0, 1.0]);
+    /// forest.update(vec![1.0, 1.0]);
+    ///
+    /// // update panics if the point has an incorrect dimensionality/length
+    /// //forest.update(vec![2.0, 3.0, 4.0]);
+    /// ```
+    pub fn update(&mut self, point: Vec<T>) {
+        assert_eq!(point.len(), self.dimension,
+            "Dimension mismatch. Expected {}-dimensional input.",
+            self.dimension);
+
+        self.num_observations += 1;
+        for tree in self.trees.iter_mut() {
+            tree.update(point.clone(), self.num_observations)
+        }
+    }
+
+    /// Return the dimension of the data accepted by this random cut forest.
+    pub fn dimension(&self) -> usize { self.dimension }
+
+    /// Return the decay factor of the random samplers used by the forest's trees.
+    pub fn time_decay(&self) -> f32 { self.time_decay }
+
+    /// Return the total number of observations made by this forest.
+    pub fn num_observations(&self) -> usize { self.num_observations }
+
+    /// Return the number of trees in this forest.
+    pub fn num_trees(&self) -> usize { self.num_trees }
+
+    /// Return the number of samples/observations stored in each tree.
+    pub fn sample_size(&self) -> usize { self.sample_size }
+
+    /// Return a vector of references to the trees of the forest.
+    pub fn trees(&self) -> &Vec<SampledTree<T>> { &self.trees }
+}
+
+
+/// Convenient mechanism for creating [`RandomCutForest`]s.
+///
+/// Random cut forests are highly configurable and come with large number of
+/// parameters, many of which have reasonable default values. This builder
+/// makes it easier to construct a random cut forest model
+///
+/// The builder has the following required parameters for initialization:
+///
+/// * `dimension`
+///
+/// The builder uses the following defaults for the remaining parameters:
+///
+/// * `num_trees = 50`
```

Comment:
Out of curiosity, is there any way in Rust where these wouldnt need to be hardcoded?